### PR TITLE
Create stdlib_modern.js

### PR DESCRIPTION
**Pull request by TyOverby**
Depends on https://github.com/ocsigen/js_of_ocaml/pull/937

* Uses the spread operator in callgen to avoid allocating more arrays
* Uses `List.slice` to implement some primitives

### CODE REVIEW

File: runtime/stdlib.js [hunk 1]
```diff
@@ -21,17 +21,12 @@
 
 //Provides: raw_array_sub
 function raw_array_sub (a,i,l) {
-  var b = new Array(l);
-  for(var j = 0; j < l; j++) b[j] = a[i+j];
-  return b
+  return a.slice(i, i + l);
```
Comment (vouillon) [inline runtime/stdlib.js]: Apparently, `slice` used to be slower (see 1bcded11b03aa2d335ef45073072994b2fc898d4). Is this still the case?

Comment (TyOverby) [reply runtime/stdlib.js]: According to [this microbenchmark](http://jsben.ch/slRuP), `slice` is three times faster

**Comment by TyOverby** [comment]
Ping @hhugo

**Comment by hhugo** [comment]
Could you add a test that would explicitly link with this modern_stdlib.js

**Comment by hhugo** [comment]
In particular, I want to make sure the file is valid

**Comment by hhugo** [comment]
Ping @TyOverby

**Comment by TyOverby** [comment]
@hhugo: I tried to write a test that passes the file through jsoo_minify just to make sure that the syntax is correct.  However, it's reporting that the file can't be parsed, but that the unparsable line is this one:

```javascript
    if (d === 0) {
//                  ^ this is character 15
```
```
../../install/default/bin/jsoo_minify: Error: error at file:"stdlib_modern.js" l:24 col:15
```

Do you know what's going on here?

**Comment by hhugo** [comment]
There is possibly a one off error when reporting the line. Could it be failing at line 25 ?

**Comment by TyOverby** [comment]
@hhugo Looks like the issue was due to me not rebasing often enough haha

Are you happy with this test that verifies syntax, or should I find a way for us to test the semantics too?

**Comment by OlivierNicole** [comment]
I’m trying to figure out what `stdlib_modern.js` is for and what is its role compared with `stdlib.js`. Would it be possible to clarify this for me @TyOverby @hhugo?

**Comment by hhugo** [comment]
I believe it was an attempt from @TyOverby to speedup part of the runtime by relying on newer features of JS.
It's used internally at janestreet.

**Comment by OlivierNicole** [comment]
I see, thanks.

**Pull request merged**